Update the contributing guidelines

floriank opened the pull request:
## Goal

In the light of some of the latest pull requests it became apparent hat we lack a single source of truth when it comes to the guidelines for contributing to OpenProject.

After some reviewing and looking into the history, I updated the guidelines for contributing to OpenProject to reflect current practices.
## Notes

Information has been pulled together from different source, mainly the [Guidelines from the OpenProject](https://community.openproject.org/projects/openproject/wiki/Git_Workflow) wiki (which is no longer available to me), comments made by @linki in #1362 and the discussion in #2594, as well as discussion with @ulferts on current practices in handling PRs.

There also was a discussion with @myabc last week on using [CLAHub](https://www.clahub.com/) for contributors to the repository, to which these guidelines have no reflection of. I'd be happy on discussing this topic, as there is some need of that in my opinion.
## Ticket

https://community.openproject.org/work_packages/18827

Commit:
updates the contributing guidelines

Update the contributing guidelines to provide a single source of truth
for contributions from the community.

Signed-off-by: Florian Kraft <f.kraft@finn.de>

floriank committed:
emphasize grace period

also, use  a different phrasing for the grace period.

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://floriank-fix-update-contributing.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://floriank-fix-update-contributing.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

Commit:
add a note on connection to work packages

[ci skip]

Signed-off-by: Florian Kraft <f.kraft@finn.de>

NobodysNightmare commented on CONTRIBUTING.md:
```diff
@@ -1,7 +1,50 @@
 OpenProject is an open source project and we encourage you to help us out. We'd be happy if you do one of these things:
```
I think you should also reword this intro or keep the list of things that someone might want to do to help us (create tickets, be active in forum, etc.).

**tl; dr:** Your write-up below does not match this paragraph

floriank replied:
Funnily enough, I had a different version which got lost in my git-fu.

I'd reword it to something along the lines of:

"If you want to help us, please read the following guidelines to contributing:"

Optionally we could reword the individual headlines to be more "pro-active", like "Creating tickets in OP", "Creating new features", "Creating hotfixes", etc.

NobodysNightmare commented on CONTRIBUTING.md:
````diff
@@ -1,7 +1,50 @@
+- Optional: Add the original OpenProject repository as a remote, so you can fetch changes: 
+
+```
+git remote add upstream git@github.com:/opf/openproject
````
I'd say the `:` after `github.com` is superfluous.

floriank replied:
I'd say I just copied it from the now unavailable git_Workflow from the OpenProject wiki page :laughing:

NobodysNightmare replied:
Better copy from GitHub:

```
https://github.com/opf/openproject.git
```

note that I obtained that link from a private tab (i.e. not being logged in, not having write-access to the repo). I think SSH URL only works with write access, but I am not sure

NobodysNightmare commented on CONTRIBUTING.md:
````diff
@@ -1,7 +1,50 @@
+git remote add upstream git@github.com:/opf/openproject
+```
+
+- Make sure you're on the right branch, right now (February 2015), the main development branch is dev: 
````
I don't think that the timestamp (February 2015) should be neccessary. This information is not allowed to out-date. We will only get uncertainty by such phrases... Better:

> Make sure you're on the right branch. The main development branch is dev

NobodysNightmare replied:
Also: What if I have something from what I believe should be included in a backport branch?

Is this possible at all? What should I do, which limitations are there? How is this backport branch called?

Azure7111 commented on CONTRIBUTING.md:
```diff
@@ -1,7 +1,50 @@
+
+- Please add tests to your code to verify functionality, especially if it is a new feature. Please also run these tests locally.
+- Please create pull requests against the current `dev` branch. Hotfixes and Bugfixes should be created against the appropiate `release/*` branch.
+- We want to keep the Pull request list as cleaned up as possible - we will close pull requests after an inactivity period of 48 hours (no comments, no further pushes)
```
Although I like the idea of closing inactive pull requests we might have to work out something more flexible here in terms of when to close a PR. Think about the holidays people usually enjoy at the end of a year, or a sudden increase in contributions. Surely there might be even more things to concern when designing something around inactivity. On the other hand this might lead to very complex rules that are difficult to understand, therefore if it works we could probably just handle this rule less strictly.

floriank replied:
I added this speciffically to avoid aging PRs.

My suggestion would to rephrase it to something along the lines of "The aim is to clouse PRs that are inactive for more than 72 hours which do not have a WIP label".

Ideally, the goal should be not to have multiple active working sites concurrently. Overall it should reduce the number of PRs, especially inactive ones.

Azure7111 replied:
That might just lead to stale PR's having a WIP label tacked on.

Commit:
remove the timestamp

The timestamp is superfluous, as the information for contributing should
not expire over time.

[ci skip]

Signed-off-by: Florian Kraft <f.kraft@finn.de>

NobodysNightmare commented on CONTRIBUTING.md:
```diff
@@ -1,7 +1,50 @@
+Ideally, there is a work package for every issue that is tackled by a contributor - the list can be found [here](https://community.openproject.org/projects/openproject/work_packages). Work packages can be created there on-demand if necessary.
+
+## Development flow
+For contributing source code, please follow the Git Workflow below:
```
We should probably add some wording to indicate, that the rules below apply to community contributions.

While the OpenProject core developers should also try to stick to these rules, they might deviate from them, e.g. not fork the repo, but have a branch on the main repository.

Proposal:

> For external contributions [...] please follow the Git Workflow below:

Azure7111 commented on CONTRIBUTING.md:
```diff
@@ -1,7 +1,46 @@
+- Please add tests to your code to verify functionality, especially if it is a new feature. Please also run these tests locally.
+- Please create pull requests against the current `dev` branch. Hotfixes and Bugfixes should be created against the appropiate `release/*` branch.
+- We want to keep the Pull request list as cleaned up as possible - we will close pull requests after an inactivity period of 48 hours (no comments, no further pushes)
+- We use OpenProject for development coordination - please have a look at [the work packages list](https://community.openproject.org/projects/openproject/work_packages) for upcoming features and reported bugs.
```
Just noticed that the linked WP list also contains test tickets. Maybe it should include a small filter for that to avoid confusion.

Commit:
increase grace period, add not on WIP label

Signed-off-by: Florian Kraft <f.kraft@finn.de>

myabc commented on CONTRIBUTING.md:
```diff
@@ -1,7 +1,46 @@
+## Development flow
+For contributing source code, please follow the Git Workflow below:
+
+- Fork Openproject on GitHub
```
Open**P**roject

Commit:
reword the first paragraph

[ci skip]

Signed-off-by: Florian Kraft <f.kraft@finn.de>

NobodysNightmare commented on CONTRIBUTING.md:
```diff
@@ -1,7 +1,46 @@
+## Important notices
+
+- Please add tests to your code to verify functionality, especially if it is a new feature. Please also run these tests locally.
+- Please create pull requests against the current `dev` branch. Hotfixes and Bugfixes should be created against the appropiate `release/*` branch.
```
I think the information around non-dev PRs is incomplete:
- How do I find out which branch is the right one?
- Which kinds of PRs are allowed there (e.g. no new features)
- Should I make two PRs in such a case? (one against `release/x` and one against `dev`?)
  - the answer is no... but I can't see it ^^

floriank replied:
You have to help me out here. My understanding after a couple of weeks:
1. Features/gem updates go against `dev`
2. Hotfixes go against latest `release/*`
3. Bugfixes go into `dev`
4. Backports (bugfixes/hotfixes) go into `release/*`

Azure7111 commented on CONTRIBUTING.md:
```diff
@@ -1,7 +1,46 @@
+
+- Please add tests to your code to verify functionality, especially if it is a new feature. Please also run these tests locally.
+- Please create pull requests against the current `dev` branch. Hotfixes and Bugfixes should be created against the appropiate `release/*` branch.
+- We want to keep the Pull request list as cleaned up as possible - we will aim close pull requests after an **inactivity period of 72 hours** (no comments, no further pushes) which are not labelled as WIP.
```
Suggestion: Change `WIP` to `work in progress` as this matches the exact wording of the corresponding label.

NobodysNightmare replied:
Also note that only contributors (**we**) are able to assign labels.

Commit:
remove typo - It's called OpenProject

[skip ci]

Signed-off-by: Florian Kraft <f.kraft@finn.de>

NobodysNightmare commented:
We require a CLA to be signed, which you do not currently mention anywhere.
I'll ask Birthe whether [CLAHub](https://www.clahub.com/) is an option for us (especially legally). Otherwise the correct way should be something like sending a letter o.O

edit: One thing that I find worrying about CLAHub is that the last activity there was two years ago...

myabc commented:
> Otherwise the correct way should be something like sending a letter o.O

Electronically-signed CLA's are used by a lot of big companies now. IANAL, but I'm pretty sure electronic signatures are also valid in our jurisdiction.

If we're required to have paper CLA's – sent my post – we might as well just write _f**\* you. we don't take contributions_ :stuck_out_tongue_winking_eye:

Commit:
explicilty state the WIP label name

[ci skip]

Signed-off-by: Florian Kraft <f.kraft@finn.de>

NobodysNightmare commented:
We are in germany here... I am not sure what qualifies as electronic signature either ^^

And what does IANAL mean?

Commit:
remove strange typo

[ci skip]

Signed-off-by: Florian Kraft <f.kraft@finn.de>

NobodysNightmare commented:
After actually **reading** the [CLA](https://www.openproject.org/wp-content/uploads/2014/09/OPF-Contributor-License-Agreement_v.2.pdf) I realized, that the CLA states how it should be transferred:

> Please complete the agreement and send it together with a detached gpg signature via email to cla@openproject.org

This should be the way then... until we decide internally, that CLAHub is also fine...

floriank commented:
I am not a lawyer. IANAL.

NobodysNightmare commented:
What about Rubocop/Hound? Do I have to care about code style and coding conventions?

floriank commented:
I would add a section for style on this for new contributions, i.e. lines you touch. 
This would include a note on hound as well as using `rubocop -a` on the files to be commited.

Commit:
split up the important notes

Also, include commentary on style guidelines and pull request reviewing.

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
add paragraph for CLA

This adds a paragraph for the CLA requirement for external contributors.

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
add a note for force pushing and cancelling builds

[ci skip]

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Azure7111 commented on CONTRIBUTING.md:
````diff
@@ -1,7 +1,102 @@
+
+```
+rubcocop -a ./path/to/file
+```
````
This will obviously not fix all of the stylecop offenses. We might want to specifically give the hint that it is useful to check manually nevertheless.

However, I'm not sure if that would be out of scope for this document and should be noted somewhere else.

floriank replied:
Well, the goal imho should be to have it all in one place, otherwise we will have the Problem of multiple documents floating around having to be cared for as well.

We could rephrase that paragraph into a suggestion that this can be done to improve the style of one's own code.

Azure7111 replied:
At the end of the day my idea of abiding code style guidelines would simply be to not introduce new offenses (no matter what kind) and maybe even go the extra mile to remove some others as well. I usually like to remove the other offenses in the files I touched in a PR. However this could really put our rubocop configuration to the test, which might not be as desirable.

NobodysNightmare replied:
I would not explicitly suggest that for the community (going the extra mile), but keep that as our mission...

This will not stop brave members of the community to fix the style nevertheless, but I think this is nothing we need to anticipate inour guidelines.

NobodysNightmare commented on CONTRIBUTING.md:
````diff
@@ -1,7 +1,102 @@
+gulp lint
+```
+
+Additionally, we use HoundCI to verifiy the styles for Ruby, as well as for JavaScript. We will reject pull requests which do not meet the code style requirements.
````
Is reject the right word here?

We use reject for PRs without description, because in doubt we want to close them quickly... I don't think we would immediately close a PR for style violations, but ask the OP to fix the style issues first.

However, I could also misunderstand the matter and we do not immediately close any PRs.

If you think I am right I would propose something along those lines:

> We will not merge pull requests which...

floriank replied:
This is correct. I'll change that wording.

Commit:
remove guidelines for code style

Signed-off-by: Florian Kraft <f.kraft@finn.de>

floriank commented:
There is a pull request for the guidelines - https://github.com/floriank/openproject/pull/2

Commit:
fix typo

[ci skip]

Signed-off-by: Florian Kraft <f.kraft@finn.de>

floriank commented:
I discussed his suggestions with @marutosi and he has a point.

I would go along with him and open this matter for discussion:
- making a workflow with a forked repository mandatory for all members of the project
- creating pull requests to be worked on by multiple members 
- setting up refspec locally so that the PRs can be inspected locally
- removing the branch restriction from the `.travis.yml`

floriank commented:
![captainplanet4](https://cloud.githubusercontent.com/assets/498241/6391950/f2ddeef4-bdbb-11e4-987c-6540ef158a11.JPG)

Summoning @ulferts @kgalli @jonasheinrich @machisuji @linki for comments.

machisuji commented:
I agree with @floriank and @marutosi on the matter of PRs through forks.
Capt. Planet would approve.

kgalli commented on CONTRIBUTING.md:
```diff
@@ -1,7 +1,82 @@
+
+## Important notes
+
+To ensure a smooth workflow for, please take note of the following:
```
I think the 'for' can be deleted. The rest sounds great so :+1:  I totally agree on PRs through forks.

floriank replied:
I think I had "for **everyone involved**" in mind.

jonasheinrich commented on CONTRIBUTING.md:
```diff
@@ -1,7 +1,82 @@
+
+Lastly, be nice and respectful to each other. We are working hard to make OpenProject the best project management software there is and we are grateful for each contribution. 
+
+If you want to get in touch with us, there is also a [Gitter channel](https://gitter.im) to talk to us directly.
```
Should we specify the Gitter channel here?

jonasheinrich commented on CONTRIBUTING.md:
```diff
@@ -1,7 +1,82 @@
+
+Please add tests to your code to verify functionality, especially if it is a new feature.
+
+Pull requests will be verified by TravisCI as well, but please run them locally as well. We have a lot of pull requests coming in and it takes some time to run the complete suite for each one.
```
Maybe we should say here that it would be best if the tests were green when run locally. So there is a chance that Travis needs to test this PR only once.

Commit:
correct typos, clarify exceptions

[ci skip]

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
specify gitter channel

[ci skip]

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
emphasize on running tests

just make sure they are green.

[ci skip]

Signed-off-by: Florian Kraft <f.kraft@finn.de>

myabc commented:
@floriank to reduce the amount of work Travis has to do, you can add `[ci skip]` to your commit messages.

floriank commented:
@myabc I think I added that - but in a more demanding way:

> Please also use `[ci skip]` in your commit message to suppress builds which are not necessary (e.g. after fixing a typo in the README).

myabc commented:
@floriank oh yes. / gitignore me

floriank commented:
```
echo "@myabc" >> .gitignore
```

Azure7111 commented:
Should contributions to plugins follow the same guidelines?

floriank commented:
Yeah, I guess so.

kgalli commented:
I strongly recommend they do!

floriank commented:
Question is - how would we redirect the users - just place a link to this `CONTRIBUTING.md` in the `CONTRIBUTING.md` in the plugin repos?

floriank commented:
I will keep this PR open till Monday - after that it will be merged.

myabc commented on CONTRIBUTING.md:
```diff
@@ -1,7 +1,84 @@
+
+### Etiquette and communication
+
+Lastly, be nice and respectful to each other. We are working hard to make OpenProject the best project management software there is and we are grateful for each contribution. 
```
@floriank Bundler now adds a Code of Conduct with its gem generator.
- https://github.com/bundler/bundler/blob/master/lib/bundler/templates/newgem/CODE_OF_CONDUCT.md.tt
- http://contributor-covenant.org/

Angular are also using it.

Might be something to consider.

floriank replied:
Thanks, I'll have a look at it after workout :muscle:

myabc commented:
:+1:

myabc commented:
Two things that just occurred to me: 
- how (or will) Teatro instances work with this workflow? 
- ditto for Rubocop.

Sorry for late feedback.

floriank commented:
Teatro and Hound should be available for all PRs - am I missing something? Rubocop can also be executed locally.

Features which would be developed by multiple developers can be created as a pull request set as `work in progress` - alternatively there is still the option to have a branch in the main repo, if desirable. the goal here should be to reduce the overall number.

myabc commented:
> Teatro and Hound should be available for all PRs - am I missing something? Rubocop can also be executed locally.

Right, of course. I was missing something.

ulferts commented on CONTRIBUTING.md:
```diff
@@ -1,7 +1,84 @@
+
+Please add tests to your code to verify functionality, especially if it is a new feature. 
+
+Pull requests will be verified by TravisCI as well, but please run them locally as well and make sure there are green before creating your pull request. We have a lot of pull requests coming in and it takes some time to run the complete suite for each one.
```
typo:

`and make sure there are green`

->

```
and make sure they are green
```

ulferts commented on CONTRIBUTING.md:
```diff
@@ -1,7 +1,84 @@
+The main development branch for upcoming releases is `dev`. For identifying the branch to create a pull request against, please refer to these rules:
+
+- If in doubt, create your pull request against `dev`. All new features, gem updates and bugfixes for the upcoming release should go into the `dev` branch.
+- Hotfixes should be created against the appropiate `release/*` branch. Backports for fixes also go against their specific `release/*` branch.
```
The term "Hotfix" is not very specific. I think we should clarify it. For example: 

> Bugfixes for one of the actively supported versions of OpenProject should be issued against the respective branch. A fix for the current version (called "Hotfix" and the branch ideally being named `hotfix/XYZ`) should target `release/*` and a fix for the former version (called "Backport" and the branch ideally being named `backport/XYZ`) should target `backport/*`. We will try to merge hotfixes into dev branch but if that is no trivial task, we might ask you to create another PR for that.

floriank replied:
:+1:

ulferts commented:
I think it's good enough. There will be stuff we might want to improve but I rather have some guidelines now that we can improve later on. 

We can issue a PR changing and following the guidelines at the same time. WHOA.  

![giphy](https://cloud.githubusercontent.com/assets/617519/6439283/935a90b8-c0d1-11e4-8386-14824ceef027.gif)

Commit:
update the branching model to be more specific

[ci skip]

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
add the Contributors Code of Conduct

[ci skip]

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
Move the Code of Conduct to it's own file

[ci skip]

Signed-off-by: Florian Kraft <f.kraft@finn.de>

floriank commented:
This is it folks. i am merging this. Thanks for all your contributions.

Here is a picture of my cat:

![img_20150117_133204](https://cloud.githubusercontent.com/assets/498241/6439486/5f87e4d2-c0d3-11e4-8272-a195b496c30f.jpg)
